Zend/zend_gl.c

```diff
@@ -0,0 +1,65 @@
+/*
+   +----------------------------------------------------------------------+
+   | PHP Version 7                                                        |
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 2.00 of the Zend license,     |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.zend.com/license/2_00.txt.                                |
+   | If you did not receive a copy of the Zend license and are unable to  |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@zend.com so we can mail you a copy immediately.              |
+   +----------------------------------------------------------------------+
+   | Authors: krakjoe@php.net                                             |
+   +----------------------------------------------------------------------+
+*/
+
+#ifndef HAVE_ZEND_GL
+#define HAVE_ZEND_GL
+
+#include "zend.h"
+#include "zend_gl.h"
+
+#include <pthread.h>
+#include <sys/mman.h>
+
+typedef pthread_mutex_t zend_gl_t;
+
+static zend_gl_t* zend_gl;
+
+int zend_gl_startup(void) {
+    pthread_mutexattr_t attributes;
+    zend_gl = mmap(NULL, sizeof(zend_gl_t),
+            PROT_READ | PROT_WRITE, MAP_SHARED | MAP_ANON, 0, 0);
+
+    if (!zend_gl) {
+        return FAILURE;
+    }
+    
+    if (pthread_mutexattr_init(&attributes) != SUCCESS ||
+        pthread_mutexattr_setpshared(&attributes, PTHREAD_PROCESS_SHARED) != SUCCESS ||
+        pthread_mutex_init(zend_gl, &attributes) != SUCCESS) {
+        munmap(zend_gl, sizeof(zend_gl_t));
+        zend_gl = NULL;
+        return FAILURE;
+    }
+
+    return SUCCESS;
+}
+
+zend_bool zend_gl_acquire() {
+    return pthread_mutex_lock(zend_gl) == SUCCESS;
```

Comment:
I never check this return, or for release ... extension code may want to, but internally we cannot halt if the call fails ... and given correct usage (which we have internally), it will not fail.